crates/sql-sink/src/main.rs

```diff
@@ -27,49 +29,75 @@ const BACKOFF_MAX: Duration = Duration::from_secs(3600 * 24);
 
 #[connector(sink)]
 async fn start(config: SqlConfig, mut stream: impl ConsumerStream) -> Result<()> {
-    let mut sink = start_sink(config).await?;
-    while let Some(item) = stream.next().await {
-        let operation: Operation = serde_json::from_slice(item?.as_ref())?;
-        trace!(?operation);
-        sink.send(operation).await?;
+    let mut backoff = backoff_init()?;
+    let mut sink = start_sink(&mut backoff, &config).await?;
+
+    info!("Starting to process records");
+
+    while let Some(item_result) = stream.next().await {
+        match item_result {
+            Ok(item) => match serde_json::from_slice::<Operation>(item.as_ref()) {
+                Ok(operation) => {
+                    trace!(?operation);
+                    while let Err(e) = sink.send(operation.clone()).await {
+                        error!("Error sending operation to sink: {}", e);
+                        sink = start_sink(&mut backoff, &config).await?;
+                    }
+                }
+                Err(err) => {
+                    error!("Failed to deserialize operation: {}", err);
+                    backoff_and_wait(&mut backoff).await?;
+                    continue;
+                }
+            },
+            Err(err) => {
+                error!("Error reading from stream: {}", err);
+                backoff_and_wait(&mut backoff).await?;
+            }
+        }
     }
 
+    info!("Connection dropped, shutting down");
+
     Ok(())
 }
 
-async fn start_sink(config: SqlConfig) -> Result<LocalBoxSink<Operation>> {
-    let mut backoff = backoff_init()?;
+async fn backoff_and_wait(backoff: &mut ExponentialBackoff) -> Result<()> {
+    let wait = backoff.wait();
+    if wait < BACKOFF_MAX {
+        warn!("Retrying in {}", humantime::format_duration(wait));
+        async_std::task::sleep(wait).await;
+        Ok(())
+    } else {
+        let err_msg = "Max retry wait time exceeded, shutting down";
+        error!(err_msg);
+        Err(anyhow!(err_msg))
+    }
+}
+
+async fn start_sink(
+    backoff: &mut ExponentialBackoff,
+    config: &SqlConfig,
```

Comment:
Backoff should be initialized

Reply:
why not initialize before and pass to it? it allows us to use the backoff in other functions